src/cluster/cluster.go

```diff
@@ -1,90 +1,123 @@
+func (c *Cluster) AddDevice(newDevice device.Device) error {
+	if viper.GetString("ENV") == "production" {
+		for _, micro := range c.Devices {
+			if micro.Host == newDevice.Host && newDevice.Port == micro.Port {
+				return errors.New("This host & port combination are already registered to this cluster")
+			}
 		}
 	}
-	return &mc.Microcontroller{}
-}
 
-//Load -
-func (c *Cluster) Load(config Config) {
-	c.Name = config.Name
-	newPeers := []mc.Microcontroller{
-		*c.Me,
-	}
-	//c.Microcontrollers = make([]mc.Microcontroller, len(config.Microcontrollers))
-	for _, micro := range config.Microcontrollers {
-		if c.Me.ID != micro.ID {
-			newPeers = append(newPeers, mc.Microcontroller{})
-			newPeers[len(newPeers)-1].Load(micro)
-		}
-	}
-	c.Microcontrollers = newPeers
+	c.Devices = append(c.Devices, newDevice)
+
+	PrintClusterInfo(*c)
+	return nil
 }
 
-func (c Cluster) String() string {
-	cluster, err := utilities.StringJSON(c)
-	if err != nil {
-		return ""
+//RemoveDevice -
+func (c *Cluster) RemoveDevice(ImDoneHere device.Device) {
+	for index, device := range c.Devices {
+		if device.ID == ImDoneHere.ID {
+			s := c.Devices
+			count := len(c.Devices)
+			s[count-1], s[index] = s[index], s[count-1]
+			c.Devices = s[:len(s)-1]
+			return
+		}
 	}
-	return cluster
 }
 
-//Start registers this microcontroller, retrieves cluster config, loads local components and verifies peers
+//Start begins the registration and health check go processes
 func (c *Cluster) Start() {
-	//Set global ref to cluster
-	gofireMaster := viper.GetBool("GOFIRE_MASTER")
-	if gofireMaster {
-		log.Println("Master Mode Enabled!")
-		c.KingMe()
-	} else {
-		log.Println("Slave Mode Enabled.")
-		c.ALifeOfServitude()
-	}
+	// Device Discovery
+	go func() {
+		ticker := time.NewTicker(30 * time.Second)
+		for {
+			select {
+			case t := <-ticker.C:
+				log.Println("Begin Device Discovery...", t)
+				for i := 1; i < 255; i++ {
+
+					host := viper.Get("SUBNET_ROOT").(string) + strconv.Itoa(i)
+					unregistered := !c.isRegistered(host)
+
+					if unregistered {
+						resp, err := http.Get("http://" + host + "/registration")
+						if err != nil || resp.StatusCode != 200 {
+							log.Println(host + " failed to respond to a registration request.")
+						} else {
+							log.Println(host + " is ready to register!")
+
+						}
+					}
+				}
+			}
+		}
+	}()
+
+	// Health Check
+	go func() {
+		ticker := time.NewTicker(10 * time.Second)
+		failCount := 0
+		failThreshold := 5
+		for {
+			select {
+			case t := <-ticker.C:
+				log.Println("Begin Heartbeat Check", t)
+				for _, m := range c.Devices {
+					log.Println("Checking Peer:", m.Name, m.ToFullAddress())
+					url := "http://" + m.ToFullAddress() + "/v1/health"
+					resp, err := http.Get(url)
+					if err != nil || resp.StatusCode != 200 {
+						log.Println(m.Name + " @" + m.ToFullAddress() + " is NOT ok")
```

Comment:
Checking if the statusCode is 404 would clear up the case in which there just is no server at that IP. Then this error message could be more descriptive